sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubConsumer.cs

```diff
@@ -28,42 +33,70 @@ public class EventHubConsumer : IAsyncDisposable
         /// <summary>The name of the default consumer group in the Event Hubs service.</summary>
         public const string DefaultConsumerGroupName = "$Default";
 
+        /// <summary>The size of event batch requested by the background publishing operation used for subscriptions.</summary>
+        private const int BackgroundPublishReceiveBatchSize = 50;
+
+        /// <summary>The maximum wait time for receiving an event batch for the background publishing operation used for subscriptions.</summary>
+        private readonly TimeSpan BackgroundPublishingWaitTime = TimeSpan.FromMilliseconds(250);
+
+        /// <summary>The set of channels for publishing events to local subscribers.</summary>
+        private readonly ConcurrentDictionary<Guid, Channel<EventData>> ActiveChannels = new ConcurrentDictionary<Guid, Channel<EventData>>();
```

Comment:
Considering adding and removing don't need fast indexing concurrentbag should be appropriate.

Reply:
I gave it a look, but decided against using it at this point.   I really don't like the result of having to rebuild the collection on unsubscribe.  That said, Iv'e got a note taken to revisit this next week when I'm digging into some of the deferred feedback.

Reply:
> I really don't like the result of having to rebuild the collection on unsubscribe. 

Why do you have to rebuild collection on unsubscribe when using ConcurrentBag?

Reply:
Because I'm an idiot and didn't see a `Remove`, completely missing `TryTake`.    So, yeah...  that's my bad.

Reply:
Turns out that I'm not such an idiot; I misread the `TryTake` initially as allowing a specific item to be requested.   There's no way to ask for a specific reference, only add something or take a "random" item.   So to remove something, you'll have to loop over and `TryTake` until you get the one that you want.   

But, you can't add those that you don't back in as you go, otherwise you've got no guarantee that you'll not keep getting those back.   The common pattern seems to be to start taking everything out until you find what you want and either move the others to a new collection or wait until you find your target and then add the rest back.

Given that, I'm going to stick with my earlier thought that I don't want to rebuild the collection each time I remove a subscription and stick with the dictionary, unless you've got some thoughts that I'm overlooking...

https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1?view=netcore-3.0

Reply:
You are right, ConcurrentDictionary works better here.